rustdoc_json: conversion cleanups

nnethercote opened the pull request:
A bunch of clean-to-types conversion cleanups I found while working on perf-related stuff in rustdoc_json.

r? @aDotInTheVoid

nnethercote committed:
Use `FromClean` more.

The `FromClean` trait is used a lot for converting to rustdoc-json
format. But it's not used universally; there are still some ad hoc
functions and methods for converting. This commit fixes this
inconsistency by using `FromClean` more.

The commit also introduces `FromClean` for `Box` and `Option`. This lets
a lot of `as_ref` and `map` calls be removed in favour of simple
`into_json` calls.

nnethercote committed:
Use `sym::asterisk` to avoid a `Symbol::intern` call.

nnethercote committed:
Remove some dead code.

We currently have both `FromClean<clean::Constant> for Constant` and
`FromClean<clean::ConstantKind> for Constant` which are basically
identical, but the former is unused.

nnethercote committed:
Remove some code.

It's just replicating exactly what is done by `<Vec<GenericParamDef> as
FromClean>::into_json`

aDotInTheVoid commented on src/librustdoc/json/conversions.rs:
```diff
@@ -652,27 +685,31 @@ impl FromClean<clean::Impl> for Impl {
-        generics: generics.into_json(renderer),
-        header: from_fn_header(&header),
-        has_body,
+impl FromClean<(&Box<clean::Function>, bool, rustc_hir::FnHeader)> for Function {
```
I don't like implementing `FromClean<(...)>`, and using a tuple for passing multiple arguments when the `clean` types don't line up with the `rustdoc_json_types` types. It's very stylistic, but I think it's just obscure and obtuse.

I don't think we need to centralize everything into happening via `FromClean`, and it's fine for there to be a few free-floating functions like this. (Expecailly as this doesn't need to get called via the "combinator" impls on Option/Box/Vec).

aDotInTheVoid commented on src/librustdoc/json/conversions.rs:
```diff
@@ -762,20 +801,21 @@ impl FromClean<clean::TypeAlias> for TypeAlias {
-            .expr
-            .map(|e| rendered_const(tcx, tcx.hir_body(e), tcx.hir_body_owner_def_id(e)))
-            .unwrap_or_default(),
+impl FromClean<(&clean::Static, rustc_hir::Safety)> for Static {
```
Ditto with regards to not liking `FromClean` impls for tuples.

aDotInTheVoid reviewed (commented):
First 3 commits look good.

4th commit broadly makes sense. I especially like loosing all the `.as_ref()` that was just noise, but needs to remove a few impls on tuples that are less clear than just having a function call for when there's not a 1:1 mapping from clean to json.

I think the last 2 commits make things worse and should be dropped. `Item` and `ItemEnum` are special as they're the "top level" of conversion, and there's special knowedge being shared about what things do and don't get converted. Also, with `clean::Item` going to `Option<Item>` it's not that sometimes a field is `None`, in the output, but it won't be present at all, which is different to every use of `FromClean`.

nnethercote committed:
Rename some methods.

- `convert_static` -> `from_clean_static`
- `from_function` -> `from_clean_function`

To match the pre-existing `from_clean_item` and `FromClean::from_clean`.

I left `JsonRenderer::convert_item` unchanged because it's a bit
different.

nnethercote commented:
Thanks for the fast review. I have dropped the last two commits and removed the `FromClean` impls using tuples. I also added one small new commit that renames a couple of conversion functions, for better naming consistency. Should be good to go now.

bors commented:
:pushpin: Commit dc8c2590afdc1037beaabe2e94f90dc0acaafcfb has been approved by `aDotInTheVoid`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=aDotInTheVoid dc8c2590afdc1037beaabe2e94f90dc0acaafcfb -->
<!-- homu: {"type":"Approved","sha":"dc8c2590afdc1037beaabe2e94f90dc0acaafcfb","approver":"aDotInTheVoid","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

nnethercote commented:
I'm going to r- this temporarily because I really want #142502 to merge (other things I'm doing depend on it) and this might conflict a tiny bit. I will r+ again once #142502 merges.

nnethercote commented:
@bors r-

nnethercote commented:
> What do you think about reinstating the commit that converts from_clean_item?

We can consider that in a follow-up. Let's get these changes merged.

@bors r=aDotInTheVoid rollup

bors commented:
:pushpin: Commit 2878e1cba320314e2abf74acbe60da49e6628827 has been approved by `aDotInTheVoid`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=aDotInTheVoid 2878e1cba320314e2abf74acbe60da49e6628827 -->
<!-- homu: {"type":"Approved","sha":"2878e1cba320314e2abf74acbe60da49e6628827","approver":"aDotInTheVoid","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
